File: validation/python/validators/validate_replay_poisoning.py
#!/usr/bin/env python3
"""FoxWhisper replay and poisoning scenario validator."""

from __future__ import annotations

import json
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, List

ROOT_DIR = Path(__file__).resolve().parents[3]
VECTORS_FILE = ROOT_DIR / "tests/common/handshake/replay_poisoning_test_vectors.json"
RESULTS_DIR = ROOT_DIR / "results"


@dataclass
class ScenarioResult:
    scenario: str
    valid: bool
    details: List[str]


class ReplayPoisoningValidator:
    """Validates replay, poisoning, and epoch-fork scenarios."""

    def __init__(self, vectors: Dict[str, Any]) -> None:
        self.vectors = vectors
        self.results: List[ScenarioResult] = []

    def run(self) -> List[ScenarioResult]:
        self._validate_replay_cases()
        self._validate_replay_boundaries()
        self._validate_poisoning_vectors()
        self._validate_epoch_forks()
        self._validate_malformed_eare()
        self._validate_anti_poisoning_rules()
        self._validate_replay_storm_profiles()
        return self.results

    def _record(self, scenario: str, valid: bool, details: List[str]) -> None:
        self.results.append(ScenarioResult(scenario=scenario, valid=valid, details=details))

    @staticmethod
    def _detect_replay(sequence_numbers: List[int], window_size: int) -> bool:
        seen: List[int] = []
        detection = False
        for seq in sequence_numbers:
            cutoff = seq - window_size
            seen = [value for value in seen if value >= cutoff]
            if seq in seen:
                detection = True
            seen.append(seq)
        return detection

    def _validate_replay_cases(self) -> None:
        section = self.vectors["replay_attack_detection"]
        window = section["window_size"]
        for case in section["test_cases"]:
            detection = self._detect_replay(case["sequence_numbers"], window)
            expected = case["expected_detection"]
            valid = detection == expected
            details = [
                f"window={window}",
                f"detected={detection}",
                f"expected={expected}",
                case.get("notes", "")
            ]
            self._record(f"replay_attack::{case['case']}", valid, details)

    def _validate_replay_boundaries(self) -> None:
        section = self.vectors["replay_window_boundaries"]
        window = section["window_size"]
        for case in section["test_cases"]:
            detection = self._detect_replay(case["sequence_numbers"], window)
            expected = case["expected_detection"]
            valid = detection == expected
            details = [
                f"window={window}",
                f"detected={detection}",
                f"expected={expected}",
                case.get("notes", "")
            ]
            self._record(f"replay_window::{case['case']}", valid, details)

    def _validate_poisoning_vectors(self) -> None:
        section = self.vectors["poisoning_injection"]
        for attack in section["attack_vectors"]:
            violations = 0
            for field in attack["malicious_fields"]:
                for key, value in field.items():
                    if key.startswith("expected_"):
                        suffix = key.split("expected_", 1)[1]
                        actual_key = f"actual_{suffix}"
                        if actual_key in field and field[actual_key] != value:
                            violations += 1
            valid = violations > 0
            details = [
                f"attack={attack['attack_name']}",
                f"violations={violations}",
                f"expected_defense={attack['expected_defense']}"
            ]
            self._record(f"poisoning::{attack['attack_name']}", valid, details)

    def _validate_epoch_forks(self) -> None:
        section = self.vectors["epoch_fork_detection"]
        for scenario in section["scenarios"]:
            timeline = scenario["timeline"]
            children: Dict[str, List[str]] = {}
            for entry in timeline:
                parent = entry["parent"]
                if parent is None:
                    continue
                children.setdefault(parent, []).append(entry["epoch_id"])
            fork_detected = any(len(nodes) > 1 for nodes in children.values())
            expected = scenario["expected_fork_detected"]
            valid = fork_detected == expected
            details = [
                f"fork_detected={fork_detected}",
                f"expected={expected}",
                f"timeline_length={len(timeline)}"
            ]
            self._record(f"epoch_fork::{scenario['scenario']}", valid, details)

    def _validate_malformed_eare(self) -> None:
        section = self.vectors["malformed_eare"]
        for record in section["records"]:
            fields = record["fields"]
            required_fields = record.get("required_fields", [])
            missing = [field for field in required_fields if field not in fields]
            hash_bytes = record.get("hash_bytes")
            if hash_bytes is None:
                hash_bytes = self._safe_hex_len(fields.get("hash"))
            min_hash_bytes = record.get("min_hash_bytes", 32)
            length_ok = hash_bytes >= min_hash_bytes
            valid = not missing and length_ok
            expected_valid = record["expected_valid"]
            details = [
                f"missing_fields={missing}",
                f"hash_bytes={hash_bytes}",
                f"min_hash_bytes={min_hash_bytes}",
                f"expected_valid={expected_valid}"
            ]
            self._record(
                f"eare::{record['record_id']}",
                valid == expected_valid,
                details,
            )

    @staticmethod
    def _safe_hex_len(value: str | None) -> int:
        if not value:
            return 0
        try:
            return len(bytes.fromhex(value))
        except ValueError:
            return 0

    def _validate_anti_poisoning_rules(self) -> None:
        section = self.vectors["anti_poisoning_rules"]
        for rule in section["rules"]:
            conditions = rule["conditions"]
            sample = rule["sample_message"]
            enforced = True
            if "max_drift" in conditions:
                drift = sample["nonce_counter"] - sample["last_nonce_counter"]
                enforced = drift <= conditions["max_drift"]
            elif conditions.get("require_binding"):
                enforced = sample.get("sender_id") == sample.get("aad_sender")
            elif conditions.get("allow_missing_aad"):
                enforced = sample.get("aad") is None
            expected = rule["expected_enforced"]
            details = [
                f"rule={rule['rule_id']}",
                f"enforced={enforced}",
                f"expected={expected}"
            ]
            self._record(f"anti_poisoning::{rule['rule_id']}", enforced == expected, details)

    def _validate_replay_storm_profiles(self) -> None:
        section = self.vectors["replay_storm_simulation"]
        window = section["window_size"]
        capacity_rate = section["capacity_per_ms"]
        tolerance = 0.1
        for profile in section["profiles"]:
            burst_rate = profile["burst_rate"]
            duration = profile["duration_ms"]
            total_msgs = burst_rate * duration
            capacity = capacity_rate * duration + window
            drops = max(0.0, total_msgs - capacity)
            drop_ratio = 0.0 if total_msgs == 0 else min(1.0, drops / total_msgs)
            expected_ratio = profile["expected_drop_ratio"]
            valid = abs(drop_ratio - expected_ratio) <= tolerance
            details = [
                f"window={window}",
                f"drop_ratio={drop_ratio:.2f}",
                f"expected_ratio={expected_ratio}",
                f"burst_rate={burst_rate}",
                f"duration_ms={duration}"
            ]
            self._record(f"replay_storm::{profile['profile_id']}", valid, details)


def load_vectors(path: Path = VECTORS_FILE) -> Dict[str, Any]:
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def save_results(results: List[ScenarioResult]) -> Path:
    RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    output_path = RESULTS_DIR / "replay_poisoning_validation_results.json"
    payload = {
        "scenario_count": len(results),
        "results": [asdict(result) for result in results],
        "success": all(result.valid for result in results),
    }
    with output_path.open("w", encoding="utf-8") as handle:
        json.dump(payload, handle, indent=2)
    return output_path


def main() -> None:
    print("FoxWhisper Replay & Poisoning Validator")
    print("=" * 55)
    vectors = load_vectors()
    validator = ReplayPoisoningValidator(vectors)
    results = validator.run()

    success_count = sum(1 for result in results if result.valid)
    print(f"Validated {len(results)} scenarios: {success_count} passed")
    for result in results:
        status = "✅" if result.valid else "❌"
        print(f"{status} {result.scenario}")

    output_path = save_results(results)
    print(f"\n📄 Results saved to {output_path}")

    if any(not result.valid for result in results):
        raise SystemExit("Replay/poisoning validation failed")


if __name__ == "__main__":
    main()

File: validation/nodejs/validators/validate_replay_poisoning.js
#!/usr/bin/env node
/**
 * FoxWhisper replay & poisoning validator (Node.js)
 */

const fs = require('fs');
const path = require('path');

const { writeJson, inputPath } = require('../util/reporting');

class ReplayPoisoningValidator {
  constructor(vectors) {
    this.vectors = vectors;
    this.results = [];
  }

  run() {
    this.validateReplayCases();
    this.validateReplayBoundaries();
    this.validatePoisoningVectors();
    this.validateEpochForks();
    this.validateMalformedEare();
    this.validateAntiPoisoningRules();
    this.validateReplayStormProfiles();
    return this.results;
  }

  record(scenario, valid, details) {
    this.results.push({ scenario, valid, details });
  }

  detectReplay(sequenceNumbers, windowSize) {
    const seen = [];
    let detection = false;
    for (const seq of sequenceNumbers) {
      const cutoff = seq - windowSize;
      for (let i = seen.length - 1; i >= 0; i--) {
        if (seen[i] < cutoff) {
          seen.splice(i, 1);
        }
      }
      if (seen.includes(seq)) {
        detection = true;
      }
      seen.push(seq);
    }
    return detection;
  }

  validateReplayCases() {
    const section = this.vectors.replay_attack_detection;
    const windowSize = section.window_size;
    for (const test of section.test_cases) {
      const detected = this.detectReplay(test.sequence_numbers, windowSize);
      const expected = test.expected_detection;
      this.record(
        `replay_attack::${test.case}`,
        detected === expected,
        [
          `window=${windowSize}`,
          `detected=${detected}`,
          `expected=${expected}`,
          test.notes || '',
        ].filter(Boolean)
      );
    }
  }

  validateReplayBoundaries() {
    const section = this.vectors.replay_window_boundaries;
    const windowSize = section.window_size;
    for (const test of section.test_cases) {
      const detected = this.detectReplay(test.sequence_numbers, windowSize);
      const expected = test.expected_detection;
      this.record(
        `replay_window::${test.case}`,
        detected === expected,
        [
          `window=${windowSize}`,
          `detected=${detected}`,
          `expected=${expected}`,
          test.notes || '',
        ].filter(Boolean)
      );
    }
  }

  validatePoisoningVectors() {
    const section = this.vectors.poisoning_injection;
    for (const attack of section.attack_vectors) {
      let violations = 0;
      for (const field of attack.malicious_fields) {
        for (const [key, value] of Object.entries(field)) {
          if (!key.startsWith('expected_')) continue;
          const suffix = key.replace('expected_', '');
          const actualKey = `actual_${suffix}`;
          if (Object.prototype.hasOwnProperty.call(field, actualKey) && field[actualKey] !== value) {
            violations += 1;
          }
        }
      }
      this.record(
        `poisoning::${attack.attack_name}`,
        violations > 0,
        [
          `violations=${violations}`,
          `expected_defense=${attack.expected_defense}`,
        ]
      );
    }
  }

  validateEpochForks() {
    const section = this.vectors.epoch_fork_detection;
    for (const scenario of section.scenarios) {
      const children = {};
      for (const entry of scenario.timeline) {
        if (!entry.parent) continue;
        if (!children[entry.parent]) {
          children[entry.parent] = [];
        }
        children[entry.parent].push(entry.epoch_id);
      }
      const forkDetected = Object.values(children).some(nodes => nodes.length > 1);
      const expected = scenario.expected_fork_detected;
      this.record(
        `epoch_fork::${scenario.scenario}`,
        forkDetected === expected,
        [
          `fork_detected=${forkDetected}`,
          `expected=${expected}`,
          `timeline_length=${scenario.timeline.length}`,
        ]
      );
    }
  }

  validateMalformedEare() {
    const section = this.vectors.malformed_eare;
    for (const record of section.records) {
      const fields = record.fields || {};
      const required = record.required_fields || [];
      const missing = required.filter(field => !(field in fields));
      const hashBytes = record.hash_bytes ?? this.hexLength(fields.hash);
      const minHashBytes = record.min_hash_bytes ?? 32;
      const lengthOk = hashBytes >= minHashBytes;
      const valid = missing.length === 0 && lengthOk;
      const expected = record.expected_valid;
      this.record(
        `eare::${record.record_id}`,
        valid === expected,
        [
          `missing_fields=${missing.join(',')}`,
          `hash_bytes=${hashBytes}`,
          `min_hash_bytes=${minHashBytes}`,
          `expected_valid=${expected}`,
        ]
      );
    }
  }

  validateAntiPoisoningRules() {
    const section = this.vectors.anti_poisoning_rules;
    for (const rule of section.rules) {
      const sample = rule.sample_message || {};
      const conditions = rule.conditions || {};
      let enforced = true;

      if (typeof conditions.max_drift === 'number') {
        const drift = sample.nonce_counter - sample.last_nonce_counter;
        enforced = drift <= conditions.max_drift;
      } else if (conditions.require_binding) {
        enforced = sample.sender_id === sample.aad_sender;
      } else if (conditions.allow_missing_aad) {
        enforced = sample.aad === null || sample.aad === undefined;
      }

      this.record(
        `anti_poisoning::${rule.rule_id}`,
        enforced === rule.expected_enforced,
        [
          `enforced=${enforced}`,
          `expected=${rule.expected_enforced}`,
        ]
      );
    }
  }

  validateReplayStormProfiles() {
    const section = this.vectors.replay_storm_simulation;
    const windowSize = section.window_size;
    const capacityRate = section.capacity_per_ms;
    const tolerance = 0.1;

    for (const profile of section.profiles) {
      const total = profile.burst_rate * profile.duration_ms;
      const capacity = capacityRate * profile.duration_ms + windowSize;
      const drops = Math.max(0, total - capacity);
      const dropRatio = total === 0 ? 0 : Math.min(1, drops / total);
      const expectedRatio = profile.expected_drop_ratio;
      const valid = Math.abs(dropRatio - expectedRatio) <= tolerance;
      this.record(
        `replay_storm::${profile.profile_id}`,
        valid,
        [
          `window=${windowSize}`,
          `drop_ratio=${dropRatio.toFixed(2)}`,
          `expected_ratio=${expectedRatio}`,
          `burst_rate=${profile.burst_rate}`,
          `duration_ms=${profile.duration_ms}`,
        ]
      );
    }
  }

  hexLength(value) {
    if (!value || typeof value !== 'string') {
      return 0;
    }
    try {
      return Buffer.from(value, 'hex').length;
    } catch (err) {
      return 0;
    }
  }
}

function saveResults(results) {
  const payload = {
    language: 'nodejs',
    scenario_count: results.length,
    success: results.every(result => result.valid),
    results,
  };
  const outputPath = writeJson('replay_poisoning_validation_results_nodejs.json', payload);
  console.log(`\n📄 Results saved to ${outputPath}`);
}

function main() {
  const [, , vectorsPath] = process.argv;
  if (!vectorsPath) {
    console.log('Usage: node validate_replay_poisoning.js <test_vectors_file>');
    process.exit(1);
  }

  const resolvedPath = path.isAbsolute(vectorsPath) ? vectorsPath : inputPath(vectorsPath);
  const data = fs.readFileSync(resolvedPath, 'utf8');
  const vectors = JSON.parse(data);

  console.log('FoxWhisper Replay & Poisoning Validator (Node.js)');
  console.log('='.repeat(55));

  const validator = new ReplayPoisoningValidator(vectors);
  const results = validator.run();

  const successes = results.filter(result => result.valid).length;
  console.log(`Validated ${results.length} scenarios: ${successes} passed`);
  results.forEach(result => {
    const status = result.valid ? '✅' : '❌';
    console.log(`${status} ${result.scenario}`);
  });

  saveResults(results);

  if (results.some(result => !result.valid)) {
    process.exitCode = 1;
  }
}

if (require.main === module) {
  main();
}
